### PR TITLE
[SPARK-5154] [PySpark] [Streaming] Kafka streaming support in Python

### PR DESCRIPTION
**Pull request by davies**
This PR brings the Python API for Spark Streaming Kafka data source.

```
    class KafkaUtils(__builtin__.object)
     |  Static methods defined here:
     |
     |  createStream(ssc, zkQuorum, groupId, topics, storageLevel=StorageLevel(True, True, False, False,
2), keyDecoder=<function utf8_decoder>, valueDecoder=<function utf8_decoder>)
     |      Create an input stream that pulls messages from a Kafka Broker.
     |
     |      :param ssc:  StreamingContext object
     |      :param zkQuorum:  Zookeeper quorum (hostname:port,hostname:port,..).
     |      :param groupId:  The group id for this consumer.
     |      :param topics:  Dict of (topic_name -> numPartitions) to consume.
     |                      Each partition is consumed in its own thread.
     |      :param storageLevel:  RDD storage level.
     |      :param keyDecoder:  A function used to decode key
     |      :param valueDecoder:  A function used to decode value
     |      :return: A DStream object
```

run the example: 

```
bin/spark-submit --driver-class-path external/kafka-assembly/target/scala-*/spark-streaming-kafka-assembly-*.jar examples/src/main/python/streaming/kafka_wordcount.py localhost:2181 test
```

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala [hunk 1]
```diff
@@ -374,49 +375,61 @@ private[spark] object PythonRDD extends Logging {
...
     // entire Spark API, I have to use this hacky approach:
+    def write(bytes: Array[Byte]) {
+      if (bytes == null) {
+        dataOut.writeInt(SpecialLengths.NULL)
+      } else {
+        dataOut.writeInt(bytes.length)
+        dataOut.write(bytes)
+      }
+    }
+    def writeS(str: String) {
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala]: new line missing.

**Comment by davies** [comment]
@tdas I think this PR is almost ready, please follow the example to double check it, thanks!

**Comment by kennethmyers** [comment]
Hello,

I made a fork of the main Spark repo and pulled this pull request into it. I was able to build it successfully but I get the followng error when calling KafkaUtils.createStream( ... )

The line: 
```
kvs = KafkaUtils.createStream(streaming_context, ZOOKEEPER, GROUP, {TOPIC: 1})
```
...gives the following error message:

```
No kafka package, please put the assembly jar into classpath:
 $ bin/submit --driver-class-path external/kafka-assembly/target/scala-*/spark-streaming-kafka-assembly-*.jar
```

When I try to run that 'submit" command with the specified parameters, I get:

```
$ bin/spark-submit --driver-class-path external/kafka-assembly/target/scala-2.10/spark-streaming-kafka-assembly-1.3.0-SNAPSHOT.jar

Error: Must specify a primary resource (JAR or Python file)
Run with --help for usage help or --verbose for debug output
Using Spark's default log4j profile: org/apache/spark/log4j-defaults.properties
SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [jar:file:/home/myerske/spark/external/kafka-assembly/target/scala-2.10/spark-streaming-kafka-assembly-1.3.0-SNAPSHOT.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/home/myerske/spark/assembly/target/scala-2.10/spark-assembly-1.3.0-SNAPSHOT-hadoop1.0.4.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual binding is of type [org.slf4j.impl.Log4jLoggerFactory]
```

I've verified that the JAR file is there.

Has anybody had the same issue or can offer any help?

Thanks

**Comment by kennethmyers** [comment]
Also, I believe the warning message on line 77 of ```python/pyspark/streaming/kafka.py``` should read

 ```$ bin/spark-submit``` 

rather than

 ```$ /bin/submit```

**Comment by davies** [comment]
@kennethmyers fixed, thanks!

**Comment by davies** [comment]
@kennethmyers I think you should put the path of your scripts after spark-submit options, such as:

```
bin/spark-submit --driver-class-path xxxxx path/to/your_script
```

**Comment by kennethmyers** [comment]
@davies Thanks! That worked! I'm new to Spark and misunderstood the function of the "spark-submit" command.

File: examples/src/main/python/streaming/kafka_wordcount.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+"""
+ Counts words in UTF8 encoded, '\n' delimited text received from the network every second.
+ Usage: network_wordcount.py <zk> <topic>
+
+ To run this on your local machine, you need to setup Kafka and create a producer first
+     $ bin/zookeeper-server-start.sh config/zookeeper.properties
+     $ bin/kafka-server-start.sh config/server.properties
+     $ bin/kafka-topics.sh --create --zookeeper localhost:2181 --partitions 1 --topic test
+     $ bin/kafka-console-producer.sh --broker-list localhost:9092 --topic test
+
```
Comment (prabeesh) [inline examples/src/main/python/streaming/kafka_wordcount.py]: All the above commands  want to run from Kafka bin/ ? 
But it still create some confusion which directory we want use. In all the Spark examples bin/ refer to the bin/  of Spark.

Comment (davies) [inline examples/src/main/python/streaming/kafka_wordcount.py]: Good point, I will remove these, and put a link here.

Comment (prabeesh) [inline examples/src/main/python/streaming/kafka_wordcount.py]: All the above commands want to run from Kafka bin/ ? 
But it still create some confusion which directory we want use. In all the Spark examples bin/ refer to the bin/ of Spark.

**Comment by tdas** [comment]
@JoshRosen Can you take brief look at the python stuff?

File: core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala [hunk 3]
```diff
@@ -374,49 +375,63 @@ private[spark] object PythonRDD extends Logging {
     // The right way to implement this would be to use TypeTags to get the full
     // type of T.  Since I don't want to introduce breaking changes throughout the
     // entire Spark API, I have to use this hacky approach:
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala]: I believe that these three comment lines originally referred to the code on line 381 (now line 395), where we look at the runtime type of the first element of the iterator.  For clarity's sake, I think we should move the new `write` methods above this comment so that it stays in the right spot.

File: core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala [hunk 4]
```diff
@@ -374,49 +375,63 @@ private[spark] object PythonRDD extends Logging {
...
             case (key, value) =>
-              dataOut.writeInt(key.length)
-              dataOut.write(key)
-              dataOut.writeInt(value.length)
-              dataOut.write(value)
+              write(key)
+              write(value)
           }
+        // key is null
+        case (null, v:Array[Byte]) =>
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala]: Minor style nit: I think there should be a space after `v:`

File: core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala [hunk 5]
```diff
@@ -374,49 +375,63 @@ private[spark] object PythonRDD extends Logging {
...
+    def write(bytes: Array[Byte]) {
+      if (bytes == null) {
+        dataOut.writeInt(SpecialLengths.NULL)
+      } else {
+        dataOut.writeInt(bytes.length)
+        dataOut.write(bytes)
+      }
+    }
+
+    def writeS(str: String) {
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala]: Maybe call this `writeString`?  It's a little bit more verbose, but less prone to typos.

File: core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala [hunk 6]
```diff
@@ -374,49 +375,63 @@ private[spark] object PythonRDD extends Logging {
     // The right way to implement this would be to use TypeTags to get the full
     // type of T.  Since I don't want to introduce breaking changes throughout the
     // entire Spark API, I have to use this hacky approach:
+    def write(bytes: Array[Byte]) {
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala]: And similarly, maybe call this writeBytes?

File: python/pyspark/serializers.py [hunk 7]
```diff
@@ -484,6 +487,8 @@ def loads(self, stream):
         length = read_int(stream)
         if length == SpecialLengths.END_OF_DATA_SECTION:
             raise EOFError
+        if length == SpecialLengths.NULL:
```
Comment (JoshRosen) [inline python/pyspark/serializers.py]: This `if` could be an `elif`, I guess.

File: make-distribution.sh [hunk 8]
```diff
@@ -188,6 +188,7 @@ echo "Build flags: $@" >> "$DISTDIR/RELEASE"
 # Copy jars
 cp "$SPARK_HOME"/assembly/target/scala*/*assembly*hadoop*.jar "$DISTDIR/lib/"
 cp "$SPARK_HOME"/examples/target/scala*/spark-examples*.jar "$DISTDIR/lib/"
+cp "$SPARK_HOME"/external/kafka/scala*/*kafka*assembly*.jar "$DISTDIR/lib/"
```
Comment (pwendell) [inline make-distribution.sh]: Rather than packaging this with the release, can we just ask users to add the maven coordinates when launching it. This will add a fairly large amount to the binary size of Spark (especially if we add other ones in the future).

Comment (pwendell) [inline make-distribution.sh]: I'm assuming that #4215 gets merged in all this.

**Comment by JoshRosen** [comment]
Should we have tests for this?  Do we have tests for the other Python streaming sources?

File: python/pyspark/streaming/kafka.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+            return ssc._jvm.org.apache.spark.util.Utils.classForName(name)
+
+        try:
+            array = getClassByName("[B")
+            decoder = getClassByName("kafka.serializer.DefaultDecoder")
+            jstream = ssc._jvm.KafkaUtils.createStream(ssc._jssc, array, array, decoder, decoder,
+                                                       jparam, jtopics, jlevel)
+        except Py4JError, e:
+            # TODO: use --jar once it also work on driver
+            if not e.message or 'call a package' in e.message:
```
Comment (JoshRosen) [inline python/pyspark/streaming/kafka.py]: This is clever; the 'call a package' errors are _really_ confusing to users, so this message is pretty helpful.

File: core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala (hunk 4 above)
Comment (tdas) [inline core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala]: nit: Also, for consistency with other "cases", v --> value.

File: core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala [hunk 10]
```diff
@@ -313,6 +313,7 @@ private object SpecialLengths {
   val PYTHON_EXCEPTION_THROWN = -2
   val TIMING_DATA = -3
   val END_OF_STREAM = -4
+  val NULL = -5
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala]: So this patch tries to fix a bug in Python regarding null values? If so, that probably should be a different patch from this Kafka patch.

File: make-distribution.sh (hunk 8 above)
Comment (tdas) [inline make-distribution.sh]: I not inclined to block this PR on #4215 . We can make the doc fix separately later.

File: core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala (hunk 10 above)
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala]: @tdas I think that this same null-handling change has been proposed before but until now I don't think we had a great reason to pull it in, since none of our internal APIs relied on it and we were worried that it might mask the presence of bugs.  Now that we have a need for it, though, it might be okay to pull in here.

File: make-distribution.sh (hunk 8 above)
Comment (davies) [inline make-distribution.sh]: The motivation for the assembly jars is to simplify the process for Python programmers, can #4215  help in this case?

File: core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala (hunk 3 above)
Comment (davies) [inline core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala]: fixed

File: make-distribution.sh (hunk 8 above)
Comment (pwendell) [inline make-distribution.sh]: We will publish the jar to maven central. It's fine IMO in this patch to recommend downloading the published jar. And later on we can just suggest linking against the maven coordinates. But I'd like to avoid actually including this JAR in the binary distribution.

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

**Comment by tdas** [comment]
Jenkins, test this please.

File: core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala (hunk 10 above)
Comment (tdas) [inline core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala]: @JoshRosen I am more or less okay with this, I just fear any regressions in core pyspark that this would cause. I dont understand the implications well enough. @JoshRosen Do you have any thoughts on this?

File: external/kafka-assembly/pom.xml [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+          <goals>
+            <goal>shade</goal>
+          </goals>
+          <configuration>
+            <transformers>
+              <transformer implementation="org.apache.maven.plugins.shade.resource.ServicesResourceTransformer"/>
+              <transformer implementation="org.apache.maven.plugins.shade.resource.AppendingTransformer">
+                <resource>reference.conf</resource>
+              </transformer>
+              <transformer implementation="org.apache.maven.plugins.shade.resource.DontIncludeResourceTransformer">
```
Comment (tdas) [inline external/kafka-assembly/pom.xml]: @pwendell Could you do one more pass on this pom.xml

File: core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala (hunk 10 above)
Comment (tdas) [inline core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala]: Also if this is a core python change, there should be unit tests that capture this change.

Comment (davies) [inline core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala]: Good point, I will do that.

File: python/pyspark/streaming/kafka.py [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+        """
+        Create an input stream that pulls messages from a Kafka Broker.
+
+        :param ssc:  StreamingContext object
+        :param zkQuorum:  Zookeeper quorum (hostname:port,hostname:port,..).
+        :param groupId:  The group id for this consumer.
+        :param topics:  Dict of (topic_name -> numPartitions) to consume.
+                        Each partition is consumed in its own thread.
+        :param storageLevel:  RDD storage level.
+        :param keyDecoder:  A function used to decode key
```
Comment (tdas) [inline python/pyspark/streaming/kafka.py]: Shouldnt we mention what the default decoder is?

File: python/pyspark/streaming/kafka.py [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+        :param valueDecoder:  A function used to decode value
+        :return: A DStream object
+        """
+        java_import(ssc._jvm, "org.apache.spark.streaming.kafka.KafkaUtils")
+
+        param = {
+            "zookeeper.connect": zkQuorum,
+            "group.id": groupId,
+            "zookeeper.connection.timeout.ms": "10000",
```
Comment (tdas) [inline python/pyspark/streaming/kafka.py]: Why is this automatically added?

File: python/pyspark/streaming/kafka.py [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+
+def utf8_decoder(s):
+    """ Decode the unicode as UTF-8 """
+    return s and s.decode('utf-8')
+
+
+class KafkaUtils(object):
+
+    @staticmethod
+    def createStream(ssc, zkQuorum, groupId, topics,
```
Comment (tdas) [inline python/pyspark/streaming/kafka.py]: I think we should have a interface where arbitrary kafka parameters can be added.

File: core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala (hunk 10 above)
Comment (tdas) [inline core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala]: Offline conversation: This change is related to the SPARK-1630 which was attempted to be solved in #1551 Since that PR was rejected by @mateiz its better that he takes a look and comment on the validity of this patch.

File: python/pyspark/streaming/kafka.py (hunk 13 above)
Comment (tdas) [inline python/pyspark/streaming/kafka.py]: Ignore this. This is also present in the scala code, so let it be here.

File: core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala (hunk 10 above)
Comment (mateiz) [inline core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala]: So what was the discussion here? Before, the motivation was that Nones from Python would be encoded as a pickled version of None. Is it that we now want to send nulls back, and we are not using the pickle library for Java that would map them? Then it would be okay.

Comment (davies) [inline core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala]: For Kafka, each record is (K,V), but K could be null if not provided by producer. We pass the bytes from JVM to Python, do the deserialization in Python, no pickling needed. So we need some way to tell the key is null or empty string.

File: core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala [hunk 15]
```diff
@@ -371,54 +372,28 @@ private[spark] object PythonRDD extends Logging {
...
-              dataOut.write(value)
-          }
-        case other =>
-          throw new SparkException("Unexpected element type " + first.getClass)
-      }
+
+    def write(obj: Any): Unit = obj match {
+      case null =>
+        dataOut.writeInt(SpecialLengths.NULL)
+
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala]: Why the extra line?

File: core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala [hunk 16]
```diff
@@ -371,54 +372,28 @@ private[spark] object PythonRDD extends Logging {
...
+      case arr: Array[Byte] =>
+        dataOut.writeInt(arr.length)
+        dataOut.write(arr)
+      case str: String =>
+        writeUTF(str, dataOut)
+
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala]: extra line?

File: core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala [hunk 17]
```diff
@@ -371,54 +372,28 @@ private[spark] object PythonRDD extends Logging {
...
+      case stream: PortableDataStream =>
+        write(stream.toArray())
+      case (key, value) =>
+        write(key)
+        write(value)
+
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala]: extra line?

File: core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala [hunk 18]
```diff
@@ -371,54 +372,28 @@ private[spark] object PythonRDD extends Logging {
...
-              dataOut.writeInt(key.length)
-              dataOut.write(key)
-              dataOut.writeInt(value.length)
-              dataOut.write(value)
-          }
-        case other =>
-          throw new SparkException("Unexpected element type " + first.getClass)
-      }
+
+    def write(obj: Any): Unit = obj match {
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala]: nit: this method can be made a final method of scala and use tail-recursion optimization.
http://tech.pro/blog/2112/scala-tail-recursion-optimisation-and-comparison-to-java
might help in performance of large nested objects. this is can be a different PR completely.

File: core/src/test/scala/org/apache/spark/api/python/PythonRDDSuite.scala [hunk 19]
```diff
@@ -23,11 +23,21 @@ import org.scalatest.FunSuite
...
+    val input: List[String] = List("a"*100000)
+    val buffer = new DataOutputStream(new ByteArrayOutputStream)
+    PythonRDD.writeIteratorToStream(input.iterator, buffer)
+  }
 
-}
+  test("Handle nulls gracefully") {
+    val buffer = new DataOutputStream(new ByteArrayOutputStream)
+    PythonRDD.writeIteratorToStream(List("a", null).iterator, buffer)
+    PythonRDD.writeIteratorToStream(List(null, "a").iterator, buffer)
```
Comment (tdas) [inline core/src/test/scala/org/apache/spark/api/python/PythonRDDSuite.scala]: nit: List(...).iterator ---> Iterator(....)

File: core/src/test/scala/org/apache/spark/api/python/PythonRDDSuite.scala [hunk 20]
```diff
@@ -23,11 +23,21 @@ import org.scalatest.FunSuite
...
+    PythonRDD.writeIteratorToStream(List("a".getBytes, null).iterator, buffer)
+    PythonRDD.writeIteratorToStream(List(null, "a".getBytes).iterator, buffer)
 
+    PythonRDD.writeIteratorToStream(List((null, null), ("a", null), (null, "b")).iterator, buffer)
```
Comment (tdas) [inline core/src/test/scala/org/apache/spark/api/python/PythonRDDSuite.scala]: There are not assert/checks whether the data was written correctly?

File: core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala (hunk 15 above)
Comment (davies) [inline core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala]: group them into different categories.

File: core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala (hunk 18 above)
Comment (davies) [inline core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala]: Thanks! In this particular case, it only have max 2 depths.

Comment (tdas) [inline core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala]: I thought it can be as deep as the number of nesting in tuples. Like (1, (2, (3, 3))) is 3 levels deep. Yeah, thats probably not a very general usecase.

Comment (davies) [inline core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala]: This JavaRDD could only be String, Array[Byte], Stream, or a pair of them.

File: core/src/test/scala/org/apache/spark/api/python/PythonRDDSuite.scala (hunk 20 above)
Comment (tdas) [inline core/src/test/scala/org/apache/spark/api/python/PythonRDDSuite.scala]: This issue still has not been addressed. There are not asserts to check whether the nulls can be read back properly.

Comment (davies) [inline core/src/test/scala/org/apache/spark/api/python/PythonRDDSuite.scala]: This is a test in Python to verify that (but not cover all the cases).

File: core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala (hunk 10 above)
Comment (tdas) [inline core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala]: @mateiz Could take a call on this.

Comment (mateiz) [inline core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala]: Alright, it sounds good to allow nulls then.

File: external/kafka-assembly/pom.xml (hunk 11 above)
Comment (pwendell) [inline external/kafka-assembly/pom.xml]: Yes - this will be published by default (unless we explicitly change something).

File: make-distribution.sh (hunk 8 above)
Comment (pwendell) [inline make-distribution.sh]: @tdas @davies - so can we remove this then? We can assume that we are publishing the assembly jar to maven (so that they can just add it with a flag to spark submit).

Comment (tdas) [inline make-distribution.sh]: Yes, we can remove this. @davies can you please update the PR. Small change.

**Comment by davies** [comment]
@tdas  @pwendell done

File: python/pyspark/streaming/kafka.py [hunk 21]
```diff
@@ -0,0 +1,83 @@
...
+            decoder = getClassByName("kafka.serializer.DefaultDecoder")
+            jstream = ssc._jvm.KafkaUtils.createStream(ssc._jssc, array, array, decoder, decoder,
+                                                       jparam, jtopics, jlevel)
+        except Py4JError, e:
+            # TODO: use --jar once it also work on driver
+            if not e.message or 'call a package' in e.message:
+                print "No kafka package, please put the assembly jar into classpath:"
+                print " $ bin/spark-submit --driver-class-path external/kafka-assembly/target/" + \
+                      "scala-*/spark-streaming-kafka-assembly-*.jar"
+            raise e
```
Comment (tdas) [inline python/pyspark/streaming/kafka.py]: The message that gets printed here is quite scary. 
```

2015-02-02 18:31:31.950 java[76691:5f03] Unable to load realm info from SCDynamicStore
No kafka package, please put the assembly jar into classpath:
 $ bin/spark-submit --driver-class-path external/kafka-assembly/target/scala-*/spark-streaming-kafka-assembly-*.jar
Traceback (most recent call last):
  File "/Users/tdas/Projects/Spark/spark/examples/src/main/python/streaming/kafka_wordcount.py", line 46, in <module>
    kvs = KafkaUtils.createStream(ssc, zkQuorum, "spark-streaming-consumer", {topic: 1})
  File "/Users/tdas/Projects/Spark/spark/python/pyspark/streaming/kafka.py", line 80, in createStream
    raise e
py4j.protocol.Py4JJavaError: An error occurred while calling z:org.apache.spark.util.Utils.classForName.
: java.lang.ClassNotFoundException: kafka.serializer.DefaultDecoder
	at java.net.URLClassLoader$1.run(URLClassLoader.java:366)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:355)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.net.URLClassLoader.findClass(URLClassLoader.java:354)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:425)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:358)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:270)
	at org.apache.spark.util.Utils$.classForName(Utils.scala:153)
	at org.apache.spark.util.Utils.classForName(Utils.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:231)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:379)
	at py4j.Gateway.invoke(Gateway.java:259)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:133)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.GatewayConnection.run(GatewayConnection.java:207)
	at java.lang.Thread.run(Thread.java:744)

```

Its easy to miss the real message. Is it possible to quit in a such that the whole stack trace does not get printed. Rather it gracefully quits after printing this message? Perhaps a quit? @JoshRosen Is that good idea.

**Comment by tdas** [comment]
I am merging despite a small comment from me.
